### PR TITLE
Make router planner active at all times

### LINKED ISSUE #930: Always pick router executor over task tracker executor
Citus currently has three different executors:
1. Router executor: Queries that we can push down to a worker node
2. Real-time executor: Queries that span across worker nodes -- queries that touch usually touch less than ~500 shards and don't require repartitioning
3. Task-tracker executor: Queries that span across worker nodes -- queries that touch many shards or require repartitioning

Router executor offers full SQL coverage and it's also the most resource efficient.

We currently allow our users to pick between `real-time` and `task-tracker` executor types through the [citus.task_executor_type](http://docs.citusdata.com/en/v5.2/reference/configuration.html#citus-task-executor-type-enum). If we can plan a query for router execution, we know that it will always perform better than the other executors. We therefore automatically pick it over the `real-time` executor.

We however don't exhibit this behavior when the `task-tracker` is configured and this leads to inconsistent behavior. If we can push down the query to a worker node, we should always switch to the `router` executor.

### PR DESCRIPTION
**Pull request by mtuncer**
We used to block router planner if task executor is in task-tracker mode.  This really did not have much benefit other than serving as a switch to disable router planner.

This change removes blocking of router planner when in task-tracker mode. It also makes corresponding changes in regression tests. Since we have a uniform behavior between executor modes we can remove alternate test result file too.

Fixes #930

### CODE REVIEW

**Comment by anarazel** [comment]
I think we should have a specific GUC for disabling router for SELECT queries (not for modify, since there's no choice available anyway).  Right now you need to resort to ugly things to force it to off, which is both bad for us during development, and for users if they have a distinct need to force router to off (e.g. due to bugs).  @ozgune What do you think?

**Comment by onderkalaci** [comment]
As far as I can see,  this PR should also be removing the check from [here](https://github.com/citusdata/citus/blob/fa8d39ec918dfbf3209584114de09d483beadb59/src/backend/distributed/planner/multi_physical_planner.c#L281). (And cascading to header, `#include "distributed/multi_server_executor.h"` from the same file.)

(With `INSERT ... SELECT` we had to move the decision of router `executableness` of a query to the planner instead of the executor. So, there _might_ also be other places we need to update.)

**Comment by mtuncer** [comment]
@anarazel went ahead and added ```enable_router_planner``` with default value true.

@onderkalaci I think I already did that. Will double check

File: src/backend/distributed/shared_library_init.c [hunk 1]
```diff
@@ -356,6 +356,16 @@ RegisterCitusConfigVariables(void)
 		0,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
+		"citus.enable_router_planner",
```
Comment (mtuncer) [inline src/backend/distributed/shared_library_init.c]: ```enable_router_executor``` infers enabling/disabling router executor too. This change so far only deals with planner part. Do we want to control executor too ?

**Comment by onderkalaci** [comment]
> @onderkalaci I think I already did that. Will double check

I cannot see that line has changed.

**Comment by marcocitus** [comment]
This opens up the possibility for queries that previously were not passed to the logical planner to go to the logical planner again. Are we confident that we're not opening new crash cases?

**Comment by anarazel** [comment]
On 2016-12-13 08:01:13 -0800, Marco Slot wrote:
> This opens up the possibility for queries that previously were not passed to the logical planner to go to the logical planner again. Are we confident that we're not opening new crash cases?

Since one can quite easily modify queries to not fit router, that's
something possible anyway? Just write your constraint with some
superflous casts or such?  I.e. if there's such crashes, we have to deal
with them.

**Comment by ozgune** [comment]
Small note. *If* we anticipate that the user will use ``citus.enable_router_planner`` (or executor) only to resolve bugs, could we hide this GUC setting?

Our current behavior with ``citus.task_executor_type`` takes effort to document / explain to users. If we add another config setting that interacts with this one, it may take more more energy to communicate.

(I'm also personally cool with not having another guc setting, but hiding it from users also works.)

**Comment by mtuncer** [comment]
I will set ```GUC_NO_SHOW_ALL``` flag.

We still haven't answered executor part.  If there are no strong opinions, I will keep it as enable_router_planner, and let the router executor run regardless of the setting.

File: src/backend/distributed/shared_library_init.c (hunk 1 above)
Comment (anarazel) [inline src/backend/distributed/shared_library_init.c]: I'm not really following. The GUCs naming doesn't have to change behaviour.  But anyway, I don't think it's that important.

**Comment by onderkalaci** [comment]
@mtuncer I see that this PR is approved by @anarazel. But, I think we should be fixing the [line](https://github.com/citusdata/citus/blob/fa8d39ec918dfbf3209584114de09d483beadb59/src/backend/distributed/planner/multi_physical_planner.c#L281) that I've already mentioned.

The line you've changed and the one I've pointed out are the same (even the comments). That might be the reason for confusion.  The former line is in `MultiRouterPlannableQuery ()` and the latter is in `MultiPlanRouterExecutable ()`. 

(I can also write up an example case that explains the case if you want)

**Comment by mtuncer** [comment]
@onderkalaci it looks better now. 

My original intention was to limit this change to router planner only. Now we can also disable router executor.

**Comment by byucesoy** [comment]
Code looks good. Tested & confirmed that;

- On router plannable queries, we use it even if task_executor_type is task-tracker.
- If enable_router_executor is false, we do not use router executor (both for real-time and task-tracker)
- enable_router_executor is invisible to user

You can :ship: it

**Pull request merged**